File: Scripts/QueryScripts/CityDistribution.R
source("/Users/ricardocunha/Documents/FEUP/5ano/MSc\ Thesis/Thesis/Code/Scripts/Utils/mainBody.R")

cities <- vector()
citiesValues <- vector()
nTotalQueries <- 0
counter <- 0

parseString <- function(str){
  #print the 10 cities with most queries
  if(counter == 0){
    nTotalQueries <<- as.integer(str)
    counter <<- counter+1
  } else {
  #  print(str)
    aCity <- strsplit(str, separator)
    cities <<- append(cities, aCity[[1]][1])
    citiesValues <<- append(citiesValues, as.integer(aCity[[1]][2]))
  }
  
  
}

printResults <- function(filename){
  i<-1
  write("QUERIES PER CITY", file=filename, append=FALSE)
  write(paste("Total number of queries(not empty)",nTotalQueries,sep=": "), file=filename, append=TRUE)
  while(i<=length(cities)){
    write(paste(cities[i],paste(citiesValues[i],percent(citiesValues[i]/nTotalQueries),sep=" -> "),sep=" -> "),
          file=filename,append=TRUE)
    i <- i+1
  }
  
}

main <- function(){
  readFromStdin()
  filename <- paste(outputPath,"CityDistribution.txt", sep = "/")
  printResults(filename)
}

main();

File: Scripts/QueryScripts/CityDistribution.pl
#!/usr/bin/perl -w
BEGIN { push @INC, '/Users/ricardocunha/Documents/FEUP/5ano/MSc Thesis/Thesis/Code/Scripts/Utils' }
require "sessionFileIndexes.pl";
require "checkDeviceType.pl";

$totalQueries=0;
%hash = ();

sub isCityOnHash{
    $city = shift;
    return exists($hash{$city});
}

sub addCity{
    $city = shift;
    if(isCityOnHash($city)){
        $hash{$city}++;
    } else {
        $hash{$city} = 1;
    }
    
}

sub printHash{
    $counter=0;
    print "$totalQueries\n";
    foreach $key (sort{$hash{$b} <=> $hash{$a}} keys %hash){
        if($counter<20){
            print "$key>>>$hash{$key}\n";
            $counter++;
        } else {
            return;
        }
    }
}

sub main{
    while($_ = <STDIN>){
        @query = split(/>>>/, $_);
        
        if($query[$cityIndex] ne "EMPTY"){
            $totalQueries++;
            addCity($query[$cityIndex]);
        }
    }
    
    printHash;
}

main;


